contracts/UpgradeableProxy.sol

```diff
@@ -0,0 +1,124 @@
+pragma solidity ^0.4.24;
+
+/*
+ * UpgradeableProxy is the base contract for all upgradeable contracts.
+ * It implements proxy functionality, internal upgrade mechanisms, and internal
+ * pause/resume mechanisms.
+ *
+ * Implementations must handle the specific upgrade & pause rules.
+ */
+contract UpgradeableProxy {
+  event Upgraded(address indexed target);
+  event Paused();
+  event Resumed();
+
+  bytes32 private constant targetPosition = keccak256("gochain.proxy.target");
+  bytes32 private constant pausedPosition = keccak256("gochain.proxy.paused");
+
+  /*
+   * Initializes the starting target contract address. The placeholder 
+   * address is replaced during deployment to the correct address.
+   */
+  constructor() public {
+    address initialTarget = 0xEEfFEEffeEffeeFFeeffeeffeEfFeEffeEFfEeff;
```

Comment:
Maybe we should template this solidity source like the ERC contracts, and generate the byte code, rather than replacing these special byte strings.